### PR TITLE
test(integration): dynamic-stack CMA SDK sanity suite + AM 2.0 coverage + asset Content-Type fix (v1.10.1)

### PR DESCRIPTION
**Pull request by aniket-shikhare-cstk**
## Summary

Adds a comprehensive, **dynamic-stack live integration ("sanity") test suite** for the CMA Python SDK under `tests/integration/`, replacing the ~2-year-old API tests. Each run creates a fresh stack, exercises **every SDK resource method** (positive / negative / edge cases) against the live CMA API, then tears the stack down. Modeled on the JS CMA SDK sanity suite, adapted to pytest.

> **Note:** Targets `development` directly. This branch also contains PR #167's asset-scanning unit tests (already incorporated here), so **#167 can be closed as superseded** once this merges.

## What's included

**Framework (`tests/integration/framework/`)**
- `setup.py` — login → create stack → management token → (optional) Personalize project → flag-gated teardown
- `capture.py` — monkeypatches `requests` to record every call (method/url/headers/body/cURL) + transient-error retries
- `report.py` — self-contained **dashboard HTML report** (summary, coverage-by-resource, per-test request/response/cURL); honest status (passed-with-warnings, xfail)
- `helpers.py` — generators, response/error validators, `tracked_assert`, scan-status polling
- `conftest.py` — session fixtures (`ctx`, `stack`, `store`, AM stack, EICAR file), `pytest-order` sequencing, per-test header isolation, report hooks

**Coverage (`tests/integration/api/`, 31 files)** — user, org, stack, locale, environment, asset, taxonomy, terms, extension, webhook, global field, content type, label, entry, variant group, variants, entry variants, branch, alias, role, workflow, delivery/management tokens, release, release items, bulk ops, publish queue, metadata, audit log, oauth, **AM 2.0 assets**. Complex content-type schemas (modular blocks, groups, references, JSON RTE) round-tripped through entries.

**AM 2.0 / asset scanning** — covers `include_asset_scan_status=true` surfacing `_asset_scan_status` (`pending → clean | quarantined`), `am`-prefixed UIDs in DAM-enabled orgs, and the publish-only `api_version: 3.2` header. The asset-scan suite skips gracefully when the DAM-enabled org isn't configured.

**SDK bug fix (v1.10.1) — included here, see commit `fix(assets): …`**
- `Asset.update()` sent a JSON body with `Content-Type: multipart/form-data` → API 422. Fixed to `application/json`.
- `Asset.replace()` set a bare `multipart/form-data` (no boundary) → API 422. Fixed to let the HTTP layer set the boundary.
- Both verified against the live API and Contentstack's own JS SDK; version bumped `1.10.0 → 1.10.1` + CHANGELOG + unit-test assertion updated. Backward-compatible (the methods were previously always failing).

## Notes for reviewers
- `pytest-order` enforces dependency order; cross-file UIDs pass via a shared `store`.
- Legacy `tests/unit` + `tests/mock` are untouched (offline coverage retained); legacy `tests/api` is superseded by this suite (kept for now).
- Includes PR #167's asset-scanning unit tests (`tests/unit/assets/`), so the asset-scan unit coverage lands with this PR.

### CODE REVIEW

**Review by netrajpatel — DISMISSED** [review]

File: tests/integration/framework/setup.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+  3. create a fresh test stack
+  4. create a management token for it
+  5. (teardown) delete stack + logout, gated by DELETE_DYNAMIC_RESOURCES
+
+All calls go through the SDK so the request-capture layer records them. Uses the
+SDK exactly as a customer would — exercising the real artifact under test.
+
+Env vars (see .env.example):
+  Required: EMAIL, PASSWORD, HOST, ORGANIZATION
+  Optional: MFA_SECRET, DELETE_DYNAMIC_RESOURCES (default 'true')
```
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply tests/integration/framework/setup.py]: Fixed — changed the docstring to reference `tests/integration/.env` (the actual file that runners create) rather than the non-existent `.env.example`. Updated in commit 4c64ba7.

File: tests/integration/framework/setup.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+REQUIRED_ENV = ["EMAIL", "PASSWORD", "HOST", "ORGANIZATION"]
+
+
+def validate_environment() -> None:
+    missing = [k for k in REQUIRED_ENV if not os.getenv(k)]
+    if missing:
+        raise RuntimeError(
+            f"Missing required environment variables: {', '.join(missing)}. "
+            f"See tests/integration/.env.example."
+        )
```
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply tests/integration/framework/setup.py:51]: Fixed — the `RuntimeError` message now says `Create tests/integration/.env with these variables set.` instead of pointing to the missing `.env.example`. Updated in commit 4c64ba7.

File: contentstack_management/assets/assets.py [hunk 3]
```diff
@@ -174,7 +174,10 @@ def replace(self, file_path):
         """
 
         url = f"assets/{self.asset_uid}"
-        Parameter.add_header(self, "Content-Type", "multipart/form-data")
+        # Let requests build the multipart body and set Content-Type WITH a boundary.
+        # Setting a bare "multipart/form-data" (no boundary) makes the API reject the
+        # upload with 422 "Please send a valid multipart/form-data payload".
+        self.client.headers.pop("Content-Type", None)
         files = {"asset": open(f"{file_path}",'rb')}
         return self.client.put(url, headers = self.client.headers, params = self.params, files = files)
```
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply contentstack_management/assets/assets.py]: Fixed in commit 4c64ba7. `replace()` now builds a per-request header copy that excludes `Content-Type` (so `requests` can set it with the correct multipart boundary) instead of mutating the shared `client.headers` dict. The file is opened inside a `with` block, so the handle is always closed after the request regardless of success or failure.

File: contentstack_management/assets/assets.py [hunk 4]
```diff
@@ -407,7 +410,9 @@ def update(self, data):
         if self.asset_uid is None or '':
             raise Exception(ASSET_UID_REQUIRED)
         url = f"assets/{self.asset_uid}"
-        Parameter.add_header(self, "Content-Type", "multipart/form-data")
+        # Updating an asset's title/description sends a JSON body, so it must use
+        # application/json. Forcing multipart/form-data here makes the API reject
+        # the request with 422 "Please send a valid multipart/form-data payload".
         return self.client.put(url, headers = self.client.headers, params = self.params, data = data)
```
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply contentstack_management/assets/assets.py]: Fixed in commit 4c64ba7. `update()` now uses a per-request header copy with `Content-Type: application/json` set explicitly: `request_headers = {**self.client.headers, "Content-Type": "application/json"}`. This makes the call self-contained — it no longer depends on what a preceding `upload()`/`replace()` may have left in the shared headers dict.

File: tests/integration/framework/capture.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            if attempt < _TRANSIENT_RETRIES:
+                time.sleep(_RETRY_BACKOFF_SECONDS)
+            # Re-open any file handles (consumed by the failed multipart attempt).
+            files = kwargs.get("files")
+            if files:
+                for key, val in list(files.items()):
+                    name = getattr(val[1] if isinstance(val, (tuple, list)) else val, "name", None)
+                    if name and os.path.exists(name):
+                        kwargs["files"][key] = (val[0], open(name, "rb"), val[2]) if isinstance(val, (tuple, list)) else open(name, "rb")
+    raise last_exc
```
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply tests/integration/framework/capture.py:156]: Fixed in commit 4c64ba7. The retry block now:
1. Calls `fh.close()` (inside a bare `except`) before reopening, so no FD leaks between attempts.
2. Uses `(val[0], open(name, 'rb')) + tuple(val[2:])` — works for both 2-tuple `(name, fh)` and 3-tuple `(name, fh, content_type)` without crashing on a missing index.

File: tests/unit/assets/test_assets_unit.py [hunk 6]
```diff
@@ -253,4 +255,81 @@ def test_delete_folder(self):
...
+        self.assertEqual(response.request.method, "GET")
+
+    def test_upload_includes_scan_status_param(self):
+        file_path = "tests/resources/mock_assets/chaat.jpeg"
+        asset = self.client.stack(api_key).assets()
+        asset.add_param("_asset_scan_status", True)
+        response = asset.upload(file_path)
+        self.assertIn("_asset_scan_status=True", response.request.url)
+        self.assertEqual(response.request.method, "POST")
+
```
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply tests/unit/assets/test_assets_unit.py:283]: Fixed in commit 4c64ba7. All three `add_param` calls in the 'with param' tests now use `include_asset_scan_status`, `"true"` (lowercase string), matching the CMA API wire format and the integration suite. The `assertIn` checks were updated to `include_asset_scan_status=true` accordingly.

File: tests/unit/assets/test_assets_unit.py [hunk 7]
```diff
@@ -253,4 +255,81 @@ def test_delete_folder(self):
...
+    def test_scan_status_param_coexists_with_other_params(self):
+        asset = self.client.stack(api_key).assets(asset_uid)
+        asset.add_param("locale", "en-us")
+        asset.add_param("_asset_scan_status", True)
+        response = asset.fetch()
+        self.assertIn("locale=en-us", response.request.url)
+        self.assertIn("_asset_scan_status=True", response.request.url)
+        self.assertEqual(response.request.method, "GET")
```
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply tests/unit/assets/test_assets_unit.py:307]: Fixed in commit 4c64ba7. The 'without param' tests (`assertNotIn`) now check for `include_asset_scan_status` (the correct query param name) rather than `_asset_scan_status`. Also updated the coexistence test to `add_param('include_asset_scan_status', 'true')` with `assertIn('include_asset_scan_status=true', ...)`.

**Review by reeshika-h — DISMISSED** [review]

**Review by OMpawar-21 — DISMISSED** [review]
LGTM

File: contentstack_management/assets/assets.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aniket-shikhare-cstk) [reply contentstack_management/assets/assets.py]: Follow-up (commit c0ca20c): the per-request header copy approach turned out to be insufficient. `_api_client._call_request` unconditionally merges shared headers via `headers.update(self.headers)` after the per-request dict is passed in, which overwrites any keys we excluded — so `Content-Type: application/json` was still sent and the API returned 400 'Malformed Body'.

Reverted to the save/restore pattern: `saved_ct = self.client.headers.pop('Content-Type', None)` before the call, with restoration in a `finally` block so subsequent calls are unaffected. The `with open()` fix for FD safety is retained. This is the only approach that works given the SDK's internal header-merge behaviour.

**Review by reeshika-h — DISMISSED** [review]

**Review by OMpawar-21 — DISMISSED** [review]

**Review by reeshika-h — APPROVED** [review]

**Review by OMpawar-21 — APPROVED** [review]
